### ISSUE TITLE
Provide an API to check whether a currently running test/section is successful or not

### ISSUE THREAD

**Issue by eiennohito** [philsquared/Catch#1053]
## Description
When the test fails, I would like to dump the state of something non-trivial, so the usual `INFO`-based logging wouldn't work for me.

There exist a something `Catch::getResultCapture().lastAssertionPassed()`, but it is only about the last assertion, isn't it?

**Comment by horenmar** [comment]
Would [Listeners](https://github.com/philsquared/Catch/blob/master/docs/event-listeners.md#event-listeners) work for you?

**Comment by eiennohito** [comment]
Listeners would require me to build my own state and bring it around if I understand the API right. Something more lightweight would be nicer.

**Comment by horenmar** [comment]
@eiennohito Yeah, but it would likely be less work than building something with a hypothetical `lastTestPassed()` or `lastSectionPassed()`.

Consider this test case with two sections:
```cpp
TEST_CASE("foo") {
    SECTION("A superfluous section") {
        REQUIRE(false);
    }
    SECTION("A pointless section") {
        REQUIRE(true);
    }
}
```
when the first assertion fails, the rest of the section and test case is skipped (only destructors are executed). Potentially you could have a piece of code like this
```cpp
TEST_CASE("foo") {
    SECTION("A superfluous section") {
        REQUIRE(false);
    }
    if (lastSectionPassed()) {  // 
        logAllTheThings();      // <-- 
    }                           // 
    SECTION("A pointless section") {
        REQUIRE(true);
    }
}
```
but by the time a call to `logAllTheThings` happens, any state that you have prepared as part of the "superfluous section" is no longer there (since the section ended in failure). This means that you would likely end up having to perform the logging in a destructor of a guard object, but that object also has to be created last, so its destructor runs first when the section ends.

Querying last test case has a very similar problem, except more pronounced, as both less data is kept between test cases, and order of execution is less defined, meaning that there is no good place to have a logging check.

**Comment by eiennohito** [comment]
I agree on your point, but I was not talking about finding out the result of **previous** section or test. I would like to find out the result of **current** section or test.
And I would like to do the logging in the destructor of something like TestingEnvironment which basically holds the state of everything else, so it goes out of scope after all the asserts would be run. But wiring it inside the `Listener` machinery would be difficult because it lives inside the test/session scope.

**Comment by martinmoene** [comment]
If it helps, here's an [example for event listeners](https://github.com/johannesmoene/Catch/blob/examples/examples/210-Evt-EventListeners.cpp).
